pennylane/transforms/qcut.py

```diff
@@ -210,3 +213,121 @@ def tape_to_graph(tape: QuantumTape) -> MultiDiGraph:
             order += 1
 
     return graph
+
+
+# pylint: disable=too-many-branches
+def contract_tensors(
+    tensors: Sequence,
+    communication_graph: MultiDiGraph,
+    prepare_nodes: Sequence[Sequence[PrepareNode]],
+    measure_nodes: Sequence[Sequence[MeasureNode]],
+    use_opt_einsum: bool = False,
+):
+    """Contract tensors according to the edges specified in the communication graph.
+
+    This operation is differentiable. The ``prepare_nodes`` and ``measure_nodes`` arguments are both
+    sequences of size ``len(communication_graph.nodes)`` that describe the order of indices in the
+    ``tensors`` with respect to to the :class:`~.PrepareNode` and :class`~.MeasureNode` edges in the
+    communication graph.
+
+    Args:
+        tensors (Sequence): the tensors to be contracted
+        communication_graph (MultiDiGraph): the communication graph determining connectivity between
```

Comment:
```suggestion
        communication_graph (nx.MultiDiGraph): the communication graph determining connectivity between
```
(just to avoid ambiguity?)

Reply:
Question: does it matter that a communication graph is not defined? I suppose not if this is only a developer-facing function?

Reply:
Would the type hints help reduce the ambiguity? E.g., this is for the `tape_to_graph()` function:
![image](https://user-images.githubusercontent.com/49409390/152801786-b2a7c0cf-710b-41e1-a38e-f04107b219c5.png)
(as an aside, it's a shame that it automatically builds as `networkx.classes.multidigraph.MultiDiGraph` rather than `networkx.MultiDiGraph`)

Reply:
> Question: does it matter that a communication graph is not defined? I suppose not if this is only a developer-facing function?

I think it's ok at this level. I wouldn't say the function is necessarily developer-facing, but you'd have to be an advanced user to use it and should have picked up the terminology as it will have been mentioned in other functions (e.g., https://github.com/PennyLaneAI/pennylane/pull/2153) as well as (eventually) a more dedicated explanation in the `pennylane.transforms` module docs, and a demo.

Reply:
> But you'd have to be an advanced user to use it and should have picked up the terminology as it will have been mentioned in other functions 

Got it :+1:

This sounds like a good use-case for 

```rest
.. seealso:: :func:`~.other_func1`, :func:`~.other_func1`
```

If you decide to add it, this goes below the `Returns` (see https://pennylane.readthedocs.io/en/latest/code/api/pennylane.adjoint_metric_tensor.html for an example)

> Would the type hints help reduce the ambiguity? E.g., this is for the tape_to_graph() function

yep!

Reply:
It's now partly defined: f6156dbea448b4a8cca567ec02b862642a69dbb1

Reply:
> This sounds like a good use-case for

Cool, yes this would be nice to add once all the functions are in place, we'll do a second pass of the docs.